packages/flutter/lib/src/widgets/page_view.dart

```diff
@@ -264,12 +262,12 @@ class PageController extends ScrollController {
 class PageMetrics extends FixedScrollMetrics {
   /// Creates an immutable snapshot of values associated with a [PageView].
   PageMetrics({
-    @required double minScrollExtent,
-    @required double maxScrollExtent,
-    @required double pixels,
-    @required double viewportDimension,
-    @required AxisDirection axisDirection,
-    @required this.viewportFraction,
+    required double? minScrollExtent,
+    required double? maxScrollExtent,
+    required double? pixels,
+    required double? viewportDimension,
```

Comment:
these should probably be non-nullable too

Reply:
Like `pixels` there are place where null can happen. I introduced an `hasViewportDimension` but like `hasPixels` I'm not sure those names are good API compared to just check for nullability. As a user I would say that I should wrap every `.viewportDimension` into a block garded by `hasViewportDimension`.

Reply:
See da6feed457

Reply:
Another issue with non-nullable field is that `ScrollMetrics.copyWith` method doesn't work anymore if one of `has*` is false.

Reply:
I was using the same pattern for `minScrollExtent` and `maxScrollExtent` and faced this copyWith issue.

Reply:
ok let's leave all those as nullable then

Reply:
viewportDimension should get set at the same time as the scroll extents iirc, can we use the same getter for all three?